File: app/src/main/java/com/boardgamegeek/io/AuthInterceptor.java
package com.boardgamegeek.io;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.accounts.AuthenticatorException;
import android.accounts.OperationCanceledException;
import android.content.Context;
import android.text.TextUtils;

import com.boardgamegeek.auth.Authenticator;
import com.boardgamegeek.util.HttpUtils;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public class AuthInterceptor implements Interceptor {
	private final Context context;

	public AuthInterceptor(Context context) {
		this.context = context;
	}

	@Override
	public Response intercept(Chain chain) throws IOException {
		Request originalRequest = chain.request();

		AccountManager accountManager = AccountManager.get(context);
		final Account account = Authenticator.getAccount(accountManager);
		if (account != null) {
			final String authToken;
			try {
				authToken = accountManager.blockingGetAuthToken(account, Authenticator.AUTH_TOKEN_TYPE, true);
				if (!TextUtils.isEmpty(account.name) && !TextUtils.isEmpty(authToken)) {
					final String cookieValue = "bggusername=" + HttpUtils.encode(account.name) + "; bggpassword=" + authToken;
					Request request = originalRequest.newBuilder().addHeader("Cookie", cookieValue).build();
					return chain.proceed(request);
				}
			} catch (OperationCanceledException | AuthenticatorException ignored) {
			}
		}
		return chain.proceed(originalRequest);
	}

}

File: app/src/main/java/com/boardgamegeek/io/AuthInterceptor.kt
package com.boardgamegeek.io

import android.accounts.AccountManager
import android.content.Context
import com.boardgamegeek.auth.Authenticator
import com.boardgamegeek.extensions.encodeForUrl
import okhttp3.Interceptor
import okhttp3.Response
import timber.log.Timber
import java.io.IOException

class AuthInterceptor(private val context: Context?) : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val originalRequest = chain.request()
        context?.let {
            val accountManager = AccountManager.get(context)
            Authenticator.getAccount(accountManager)?.let { account ->
                try {
                    val password = accountManager.blockingGetAuthToken(account, Authenticator.AUTH_TOKEN_TYPE, true)
                    if (account.name.isNotBlank() && !password.isNullOrBlank()) {
                        return chain.proceed(
                            originalRequest.newBuilder()
                                .addHeader("Cookie", "bggusername=${account.name.encodeForUrl()}; bggpassword=$password")
                                .build()
                        )
                    }
                } catch (e: Exception) {
                    Timber.w(e)
                }
            }
        }
        return chain.proceed(originalRequest)
    }
}
